schism/palettes.c

```diff
@@ -342,11 +362,46 @@ void palette_apply(void)
 
 void palette_load_preset(int palette_index)
 {
-	if (palette_index < -1 || palette_index >= NUM_PALETTES)
+	if (palette_index < -1 || palette_index >= NUM_PALETTES || palette_index == current_palette_index)
 		return;
 
+	if (current_palette_index == -1)
+		memcpy(user_palette, current_palette, sizeof(current_palette));
+
 	current_palette_index = palette_index;
-	if (palette_index == -1) return;
-	memcpy(current_palette, palettes[palette_index].colors, sizeof(current_palette));
+
+	if (palette_index == -1) {
+		memcpy(current_palette, user_palette, sizeof(current_palette));
+	} else {
+		memcpy(current_palette, palettes[palette_index].colors, sizeof(current_palette));
+	}
+
+	cfg_save();
 }
 
+static const char palette_trans[65] = ".0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz";
+
+void palette_to_string(char *str_out) {
+	for (int n = 0; n < 48; n++)
+		str_out[n] = palette_trans[current_palette[n / 3][n % 3]];
+
+	str_out[48] = '\0';
+}
+
+int set_palette_from_string(const char *str_in) {
+	uint8_t colors[48];
+	const char *ptr;
+
+	// Remove bad characters from beginning (spaces etc.).
+	str_in = strpbrk(str_in, palette_trans);
+	if(!str_in) return 0;
+
+	for (int n = 0; n < 48; n++) {
+		if (str_in[n] == '\0' || (ptr = memchr(palette_trans, str_in[n], sizeof(palette_trans))) == NULL)
```

Comment:
since the palette translation unit is now a regular C string and not an array this can use `strchr` without the use of `sizeof(palette_trans)`